core/src/__tests__/internal/net/StitchRequestClientUnitTests.ts

```diff
@@ -0,0 +1,300 @@
+        expect(error.message).toContain("received unexpected status");
+      })
+      .then(() => {
+        // Error responses should be handled
+        when(transportMock.roundTrip(anything())).thenResolve({
+          body: "whoops",
+          headers: {},
+          statusCode: 500
+        });
+        return stitchRequestClient.doRequest(builder.build());
+      })
+      .catch((error: StitchServiceException) => {
+        expect(error.errorCode).toEqual(StitchServiceErrorCode.Unknown);
+        expect(error.message).toEqual("whoops");
+      })
+      .then(() => {
+        const headers = {
+          [Headers.CONTENT_TYPE]: ContentTypes.APPLICATION_JSON
+        };
+
+        when(transportMock.roundTrip(anything())).thenResolve({
+          body: "whoops",
+          headers,
+          statusCode: 500
+        });
+        stitchRequestClient.doRequest(builder.build());
+      })
+      .catch((error: StitchServiceException) => {
+        expect(error.errorCode).toEqual(StitchServiceErrorCode.Unknown);
+        expect(error.message).toEqual("whoops");
+      })
+      .then(() => {
+        const headers = {
+          [Headers.CONTENT_TYPE]: ContentTypes.APPLICATION_JSON
+        };
+
+        when(transportMock.roundTrip(anything())).thenResolve({
+          body: '{"error": "bad", "error_code": "InvalidSession"}',
+          headers,
+          statusCode: 500
+        });
+      })
+      .catch((error: StitchServiceException) => {
+        expect(error.errorCode).toEqual(StitchServiceErrorCode.InvalidSession);
+        expect(error.message).toEqual("bad");
+      })
+      .then(() => {
+        when(transport.roundTrip(anything())).thenReject(new StitchException());
+
+        return stitchRequestClient.doRequest(builder.build());
+      })
+      .catch((error: Error) => {
+        expect(error).toBeDefined();
+      });
+  });
+
+  it("should handle non canonical headers", () => {
+    const domain = "http://domain.com";
+    const transportMock = mock(FetchTransport);
+    const transport = instance(transportMock);
+
+    const stitchRequestClient = new StitchRequestClient(domain, transport);
+
+    // A bad response should throw an exception
+    when(transportMock.roundTrip(anything())).thenResolve({
+      headers: {},
+      statusCode: 500
+    });
+
+    const path = "/path";
+    const builder = new StitchRequest.Builder()
+      .withPath(path)
+      .withMethod(Method.GET);
+
+    const headers = {
+      [Headers.CONTENT_TYPE]: ContentTypes.APPLICATION_JSON
+    };
+
+    when(transportMock.roundTrip(anything())).thenResolve({
+      body: '{"error": "bad", "error_code": "InvalidSession"}',
+      headers,
+      statusCode: 502
+    });
+
+    expect(stitchRequestClient.doRequest(builder.build())).rejects.toEqual(
+      new StitchServiceException("bad", StitchServiceErrorCode.InvalidSession)
+    );
+
+    headers[Headers.CONTENT_TYPE_CANON] = ContentTypes.APPLICATION_JSON;
```

Comment:
I think this test is incorrect (the Java one is incorrect too). this should be something like `const canonicalHeaders = headers = { [Headers.CONTENT_TYPE]: ContentTypes.APPLICATION_JSON };`, and you should have the server return the `canonicalHeaders` for the next request. The way it's written right now, the headers contain both a `Content-Type` and a `content-type`, and the error handling code checks for the lowercase version.

As the code stands right now, I think the test would fail. You need to update `Response` so that when it is created, the headers are processed by having the keys be lowercased. See the main constructor in `Response.java` in the Java SDK.

Reply:
Looks like the preprocessing is implemented now, but the mocked request client in the test is still returning both headers. 

You can change this line to create `const canonicalHeaders = headers = { [Headers.CONTENT_TYPE]: ContentTypes.APPLICATION_JSON };`, and pass `canonicalHeaders` in line 292 and the test will be correct, and with your new headers code I think it should pass.